TestWatchStressResumeClose: watch stream goroutine leak

via https://jenkins-etcd-public.prod.coreos.systems/job/etcd-ci-ppc64/521/consoleFull

```
2017-03-22 16:03:40.140849 I | raft: 6ee3c52f7ee76689 is starting a new election at term 1
2017-03-22 16:03:40.140878 I | raft: 6ee3c52f7ee76689 became candidate at term 2
2017-03-22 16:03:40.140890 I | raft: 6ee3c52f7ee76689 received MsgVoteResp from 6ee3c52f7ee76689 at term 2
2017-03-22 16:03:40.140907 I | raft: 6ee3c52f7ee76689 became leader at term 2
2017-03-22 16:03:40.140916 I | raft: raft.node: 6ee3c52f7ee76689 elected leader 6ee3c52f7ee76689 at term 2
2017-03-22 16:03:40.141256 I | etcdserver: setting up the initial cluster version to 3.2
2017-03-22 16:03:40.141682 N | etcdserver/membership: set the initial cluster version to 3.2
2017-03-22 16:03:40.141821 I | etcdserver: published {Name:678268586749533368 ClientURLs:[unix://127.0.0.1:2131412950]} to cluster 226d745a51c7fd35
2017-03-22 16:03:40.493207 I | integration: terminating 678268586749533368 (unix://localhost:6782685867495333680)
2017-03-22 16:03:40.498630 I | integration: terminated 678268586749533368 (unix://localhost:6782685867495333680)
--- FAIL: TestWatchStressResumeClose (0.75s)
	leak.go:106: Test appears to have leaked a Transport:
		github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream.func1(0xc42010cd00, 0xc420116000, 0xc4200b7cc0)
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:577 +0x8c
		github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream(0xc4200b7cc0, 0xc420116000, 0xc4225b8840)
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:635 +0x6a0
		created by github.com/coreos/etcd/clientv3.(*watchGrpcStream).run
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:434 +0x3c8
		
		github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream.func1(0xc42010dd00, 0xc420116160, 0xc4200b7cc0)
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:577 +0x8c
		github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream(0xc4200b7cc0, 0xc420116160, 0xc4225b8840)
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:635 +0x6a0
		created by github.com/coreos/etcd/clientv3.(*watchGrpcStream).run
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:434 +0x3c8
		
		github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream.func1(0xc42010ed00, 0xc4201162c0, 0xc4200b7cc0)
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:577 +0x8c
		github.com/coreos/etcd/clientv3.(*watchGrpcStream).serveSubstream(0xc4200b7cc0, 0xc4201162c0, 0xc4225b8840)
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:635 +0x6a0
		created by github.com/coreos/etcd/clientv3.(*watchGrpcStream).run
			/var/jenkins_home/workspace/etcd-ci-ppc64/gopath/src/github.com/coreos/etcd/clientv3/watch.go:434 +0x3c8
		
[lots more of this same goroutine stack]
```

--- a/clientv3/watch.go
+++ b/clientv3/watch.go
@@ -132,6 +132,8 @@ type watchGrpcStream struct {
 	errc chan error
 	// closingc gets the watcherStream of closing watchers
 	closingc chan *watcherStream
+	// wg is Done when all substream goroutines have exited
+	wg sync.WaitGroup
 
 	// resumec closes to signal that all substreams should begin resuming
 	resumec chan struct{}
@@ -406,7 +408,7 @@ func (w *watchGrpcStream) run() {
 		for range closing {
 			w.closeSubstream(<-w.closingc)
 		}
-
+		w.wg.Wait()
 		w.owner.closeStream(w)
 	}()
 
@@ -431,6 +433,7 @@ func (w *watchGrpcStream) run() {
 			}
 
 			ws.donec = make(chan struct{})
+			w.wg.Add(1)
 			go w.serveSubstream(ws, w.resumec)
 
 			// queue up for watcher creation/resume
@@ -576,6 +579,7 @@ func (w *watchGrpcStream) serveSubstream(ws *watcherStream, resumec chan struct{
 		if !resuming {
 			w.closingc <- ws
 		}
+		w.wg.Done()
 	}()
 
 	emptyWr := &WatchResponse{}
@@ -674,6 +678,7 @@ func (w *watchGrpcStream) newWatchClient() (pb.Watch_WatchClient, error) {
 			continue
 		}
 		ws.donec = make(chan struct{})
+		w.wg.Add(1)
 		go w.serveSubstream(ws, w.resumec)
 	}
 

